cli/tests/services.bats

```diff
@@ -331,3 +331,122 @@ EOF
   run grep "process=flox_never_exit" "$_FLOX_SERVICES_LOG_FILE"
   assert_success
 }
+
+# ---------------------------------------------------------------------------- #
+
+@test "watchdog: can run klaus" {
+  run "$KLAUS_BIN" --help
+  assert_success
+}
+
+@test "watchdog: lives as long as the activation" {
+  export FLOX_FEATURES_SERVICES=true
+  setup_sleeping_services
+  run "$FLOX_BIN" activate -s -- bash <(cat <<'EOF'
+    source "${TESTS_DIR}/services/register_cleanup.sh"
+
+    log_file="$PWD/.flox/cache/$(ls .flox/cache)"
+
+    # Ensure that the watchdog is still running
+    if tail -n 1 "$log_file" | grep "exiting"; then
+      exit 1
+    fi
+EOF
+)
+  assert_success
+
+  # Ensure that the watchdog has exited now
+  log_file="$PWD/.flox/cache/$(ls .flox/cache)"
+  if ! tail -n 1 "$log_file" | grep "exiting"; then
+    exit 1
+  fi
+}
+
+@test "watchdog: exits on termination signal (SIGUSR1)" {
+  log_file=klaus.log
+  # This 'foo' is because we don't actually do anything with the registry yet
+  _FLOX_WATCHDOG_LOG_LEVEL=debug "$KLAUS_BIN" -r foo -l "$log_file" &
+  klaus_pid="$!"
+
+  # Wait for start.
+  timeout 1s bash -c "
+    while ! grep -qs 'watchdog is on duty' \"$log_file\"; do
+      sleep 0.1
+    done
+  "
+
+  # Check running.
+  run kill -s 0 "$klaus_pid"
+  assert_success
+
+  # Signal to exit.
+  run kill -s SIGUSR1 "$klaus_pid"
+  assert_success
+
+  # Wait for exit.
+  timeout 1s bash -c "
+    while kill -s 0 \"$klaus_pid\"; do
+      sleep 0.1
+    done
+  "
+}
+
+@test "watchdog: exits on shutdown signal (SIGINT)" {
+  log_file=klaus.log
+  # This 'foo' is because we don't actually do anything with the registry yet
+  _FLOX_WATCHDOG_LOG_LEVEL=debug "$KLAUS_BIN" -r foo -l "$log_file" &
+  klaus_pid="$!"
+
+  # Wait for start.
+  timeout 1s bash -c "
+    while ! grep -qs 'watchdog is on duty' \"$log_file\"; do
+      sleep 0.1
+    done
+  "
+
+  # Check running.
+  run kill -s 0 "$klaus_pid"
+  assert_success
+
+  # Signal to exit.
+  run kill -s SIGINT "$klaus_pid"
+  assert_success
+
+  # Wait for exit.
+  timeout 1s bash -c "
+    while kill -s 0 \"$klaus_pid\"; do
+      sleep 0.1
+    done
+  "
+}
+
+@test "watchdog: exits when parent doesn't match provided PID" {
+  log_file=klaus.log
+
+  # We need a test PID, but PIDs can be reused. There's also no delay on reusing
+  # PIDs, so you can't create and kill a process to use its PID during that
+  # make-believe no-reuse window. At best we can choose a random PID and skip
+  # the test if something is already using it.
+  test_pid=31415
+  if kill -0 "$test_pid"; then
+    skip "test PID is in use"
+  fi
```

Comment:
We need a PID that isn't being used. `kill -0 0` has an exit code of 0, just like a process that's currently running, so that doesn't quite work.

The idea for this test is to give the watchdog a PID that isn't running to mimic the scenario that the parent has died before calling `prctl`/`kqueue`. We have no way that I can figure out of preparing a PID and preventing it from being used, so I think the best we can do is pick a random one and check that it isn't being used.

Reply:
> The idea for this test is to give the watchdog a PID that isn't running to mimic the scenario that the parent has died before calling prctl/kqueue.

I would change the description from `exits when parent doesn't match provided PID` then.

Reply:
I'll catch that change in @dcarley 's PR